File: sdk/samples/WCFSamples/TechnologySamples/Extensibility/Rules/WCF_Router/cs/service-calculator/CalculatorService.cs
//  Copyright (c) Microsoft Corporation. All rights reserved.

using System;
using System.ServiceModel;
using System.ServiceModel.Dispatcher;

namespace Microsoft.ServiceModel.Samples
{
    [ServiceContract(Namespace = "http://Microsoft.ServiceModel.Samples/", SessionMode=SessionMode.Required)]
    public interface ICalculatorService
    {
        [OperationContract(IsOneWay = false)]
        int Add(int number);

        [OperationContract(IsOneWay = false)]
        int Subtract(int number);
    }

    [ServiceBehavior(InstanceContextMode = InstanceContextMode.PerSession)]
    public class CalculatorService : ICalculatorService
    {
        int total = 0;

        int ICalculatorService.Add(int number)
        {
            this.total += number;
            return this.total;
        }

        int ICalculatorService.Subtract(int number)
        {
            this.total -= number;
            return this.total;
        }
    }

    public class ServiceDriver
    {
        public static void Main(string[] args)
        {
            ServiceHost serviceHost = new ServiceHost(typeof(CalculatorService));
            serviceHost.Open();

            for (int i = 0; i < serviceHost.ChannelDispatchers.Count; i++)
            {
                ChannelDispatcher channelDispatcher = serviceHost.ChannelDispatchers[i] as ChannelDispatcher;
                if (channelDispatcher != null)
                {
                    for (int j = 0; j < channelDispatcher.Endpoints.Count; j++)
                    {
                        EndpointDispatcher endpointDispatcher = channelDispatcher.Endpoints[j];
                        Console.WriteLine("Listening on " + endpointDispatcher.EndpointAddress + "...");
                    }
                }
            }

            Console.WriteLine();
            Console.WriteLine("Press Enter to exit...");
            Console.ReadLine();
        }
    }
}

File: sdk/samples/WCFSamples/TechnologySamples/Scenario/Router/CPP/service-calculator/CalculatorService.cpp
// ----------------------------------------------------------------------------
// Copyright (C) 2003-2005 Microsoft Corporation, All rights reserved.
// ----------------------------------------------------------------------------

using namespace System;
using namespace System::ServiceModel;
using namespace System::ServiceModel::Dispatcher;
using namespace System::Diagnostics;
using namespace System::Configuration;

namespace Microsoft{
namespace ServiceModel{
namespace Samples{

    interface class ICalculatorService;

    [ServiceContract(Namespace = L"http://Microsoft.ServiceModel.Samples/", SessionMode=SessionMode::Required)]
    public interface class ICalculatorService
    {
        [OperationContract(IsOneWay = false)]
        int Add(int number);

        [OperationContract(IsOneWay = false)]
        int Subtract(int number);
    };


    [ServiceBehavior(InstanceContextMode = InstanceContextMode::PerSession)]
    public ref class CalculatorService : public ICalculatorService
    {
    private:

        int total;

    public:
        CalculatorService()
        {
            this->total = 0;
        }

        virtual int Add(int number)
        {
            this->total += number;
            return this->total;
        }

        virtual int Subtract(int number)
        {
            this->total -= number;
            return this->total;
        }
    };
}
}
}



int main(array<System::String ^> ^args)
{
    ServiceHost ^ serviceHost = gcnew ServiceHost( Microsoft::ServiceModel::Samples::CalculatorService::typeid );

    serviceHost->Open();


    for (int i = 0; i < serviceHost->ChannelDispatchers->Count; i++)
    {
        ChannelDispatcher ^ channelDispatcher = safe_cast<ChannelDispatcher^>( serviceHost->ChannelDispatchers[ i ] );

        for ( int j = 0; j < channelDispatcher->Endpoints->Count; j++ )
        {
            EndpointDispatcher ^ endpointDispatcher = channelDispatcher->Endpoints[ j ];
            Console::WriteLine( L"Listening on {0}...", endpointDispatcher->EndpointAddress->ToString() );
        }        
    }

    Console::WriteLine();
    Console::WriteLine(L"Press Enter to exit...");
    Console::ReadLine();
    serviceHost->Close();

    return 0;
}